### PR TITLE
Bypass encrypted fix

### PR DESCRIPTION
**Pull request by regit**
This patchset is a kind of RFC. It allows to trigger bypass early for TLS flows. I feel a bit unease with the choices so discussion is welcome.

PR builds:
- PR regit: https://buildbot.openinfosecfoundation.org/builders/regit/builds/218
- PR regit-pcap: https://buildbot.openinfosecfoundation.org/builders/regit-pcap/builds/0

### CODE REVIEW

**Comment by inliniac** [comment]
Can you explain the issue first, including how to reproduce it?

File: src/app-layer-ssl.c [hunk 1]
```diff
@@ -1213,6 +1213,13 @@ static int SSLv3Decode(uint8_t direction, SSLState *ssl_state,
                handshake must be done */
             ssl_state->flags |= SSL_AL_FLAG_HANDSHAKE_DONE;
 
+            /* Encrypted data, reassembly not asked, bypass asked, let's sacrifice
+             * heartbeat lke inspection to be able to be able to bypass the flow */
+            if ((ssl_config.no_reassemble == 1) && (StreamTcpBypassEnabled())) {
```
Comment (inliniac) [inline src/app-layer-ssl.c]: Not sure I follow the logic here. Why should this depend on bypass? Isn't the purpose of the no_reassemble option to disable tracking of ssl after the handshake?

**Comment by regit** [comment]
Regarding first commit, my thinking is the following, we should not completely bypass a flow before both end have reached the stream depth or have reached a certain state. Justification is that we need the hack to treat the other side so we can't really decide to cut only one side.

About second commit, I was doing some tests lately on local bypass, I've been surprised to see that TLS session were not seen in the flow entry as bypassed. This was due to the fact that the inspection is done till stream depth so Suricata get for example heartbeat message.

File: src/app-layer-ssl.c (hunk 1 above)
Comment (regit) [inline src/app-layer-ssl.c]: The option is disabling tracking but I was thinking that maybe some people want anyway to do packet matching. So I've chooze the most careful solution. My initial implementation was just setting the flags.

Comment (inliniac) [inline src/app-layer-ssl.c]: I reviewed the logic and I think we can do this without the bypass check.

**Comment by inliniac** [comment]
Commit message of https://github.com/inliniac/suricata/pull/2344/commits/64f1730565ada2fd026686e0d292c1f73293bc4c needs some love.

**Comment by inliniac** [comment]
Needs an update.

**Pull request closed without merging**